### PR TITLE
Add Replace.Attoparsec.Text.Lazy

### PR DESCRIPTION
**Pull request by expipiplus1**
Highly uncomfortable with the amount of duplicated code from the strict module

Still to do, tests and tests

### CODE REVIEW

**Comment by jamesdbrock** [comment]
> Highly uncomfortable with the amount of duplicated code from the strict module

I am totally comfortable with duplicated code here.

**Comment by expipiplus1** [comment]
hahaha, ok then :)

Still to do is go through the comments and make sure they're ok, add a little more explanation and some lazy text specific tests. Can try to do that later this week,

**Comment by jamesdbrock** [comment]
My feelings about duplicated code in cases like this:

When we want to generalize some code over a large or unbounded parameter space then it makes sense to parameterize the code. Parameterization means introducing abstractions. In Haskell that usually means a typeclass. 

In our case, the space of possible parameters is small and bounded. It's just these four types:

* Text
* Text.Lazy
* ByteString
* ByteString.Lazy

In my opinion it's not helpful to introduce more abstractions in situations like this. Better to copy-paste. More abstractions (typeclasses) complicate the user-facing API. Copy-paste keeps the API simple.

**Comment by expipiplus1** [comment]
Yeah, my concern is the maintenance burden. `vector` (and my `vector-sized`) suffers from the same issue. `vector` though has a generic typeclass version, and then modules reexporting all that with specific types.

I wonder if backpack could help here...

Happy to have this merged with the copypaste as is though, less work for me :)

**Comment by jamesdbrock** [comment]
I really should completely overhaul and Nixify https://github.com/jamesdbrock/replace-benchmark so that we can benchmark this properly...

**Comment by expipiplus1** [comment]
While it's nice to have this, it's actually less useful than I hoped as `streamEdit` is still strict in the entire input string as (even though the parse can't fail) attoparsec still wants to get to the end of the string before returning the `Done` result.

Shouldn't be the end of the world to change `sepCap` to just go one `sep` at a time and feed attoparsec incrementally manually

**Comment by jamesdbrock** [comment]
> While it's nice to have this, it's actually less useful than I hoped as `streamEdit` is still strict in the entire input string as (even though the parse can't fail) attoparsec still wants to get to the end of the string before returning the `Done` result.
> 
> Shouldn't be the end of the world to change `sepCap` to just go one `sep` at a time and feed attoparsec incrementally manually

The `anyTill` combinator is essentially the same as a “`sepCap` which goes one `sep` at a time”.

I actually think it wouldn't be crazy to supply a __Replace.Attoparsec.Text.Lazy__ module which only exports a lazy version of `anyTill`. `anyTill` is the most elemental function in this library, everything else can be built pretty easily from `anyTill`.

**Comment by expipiplus1** [comment]
hmm, actually it's a little tricky, because we want attoparsec to only return `Partial` when it's in the middle of a match of `sep` otherwise we'd like it to be `Done` after any amount of `anyChar`s.

**Comment by jamesdbrock** [comment]
> hmm, actually it's a little tricky, because we want attoparsec to only return `Partial` when it's in the middle of a match of `sep` otherwise we'd like it to be `Done` after any amount of `anyChar`s.

I guess you want it to be `Done` after any amount of `anyChars` so that you can throw away the unmatched `Text` because you want a constant-memory streaming parse?

I've actually never used __Data.Attoparsec.Text.Lazy__ for anything and I'm reading up to remember what it does... https://hackage.haskell.org/package/attoparsec-0.14.1/docs/Data-Attoparsec-Text-Lazy.html Apparently the `Result` type can never return `Partial`?

**Comment by jamesdbrock** [comment]
I'm getting the sense that your use case is “we’re streaming infinite data and we want to watch the stream for patterns which match this parser.” Is that right? In that case you may be interested in this library, which I have never used, but which I think is supposed to handle that case: https://hackage.haskell.org/package/pipes-parse

**Comment by jamesdbrock** [comment]
Oh, and maybe you want to edit the stream too?

**Comment by jamesdbrock** [comment]
Or maybe I'm guessing wrong, I should just ask you. What is your use case?

**Comment by expipiplus1** [comment]
I have several different edits I'd like to make over a (potentially large) lazy text and thought it'd be a bit nicer to have this lazy streaming version as some kind of fusion imitation. 

> I've actually never used Data.Attoparsec.Text.Lazy for anything and I'm reading up to remember what it does... 

It's not really all that sophisticated, it basically just drives attoparsec's incremental input with the chunks for you.

**Comment by expipiplus1** [comment]
I *think* that it should be as simple as replacing `atEnd` here with `atChunkEnd` https://github.com/jamesdbrock/replace-attoparsec/blob/48325c00aa5b21a9b0aa07eb7c632e8828a14f9b/src/Replace/Attoparsec/Text/Lazy.hs#L283 

Not sure if `atChunkEnd` really exists though...

**Comment by expipiplus1** [comment]
OK, seems to be working with lazy chunks! Basically a new implementation of `streamEditT` based on the non-chunk-spanning-on-failure `anyTill`.

**Comment by expipiplus1** [comment]
Not sure about leaving all the other (now unused) functionality like `sepCap`, `findAll[Cap]`, `breakCap` and `splitCap` as these are all strict in the input.

**Comment by expipiplus1** [comment]
hmm, it's definitely superlinear on the dense benchmark...

**Comment by expipiplus1** [comment]
```
for ((i=1;i<=5000000;i++)); \
do \
        echo -n "x " >> input/bench-test-dense.txt; \
done;
Thu Jul 29 01:05:33 PM +08 2021
PRETTY_NAME="NixOS 21.11 (Porcupine)"
model name      : Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
sed (GNU sed) 4.8
Python 3.9.5
This is perl 5, version 34, subversion 0 (v5.34.0) built for x86_64-linux-thread-multi
sed                      dense  401.65,msec,task-clock:u,401645998,100.00,0.998,CPUs utilized
python3                  dense  272.97,msec,task-clock:u,272967507,100.00,0.998,CPUs utilized
python3 function         dense  578.65,msec,task-clock:u,578649156,100.00,0.999,CPUs utilized
perl5                    dense  184.33,msec,task-clock:u,184327154,100.00,0.995,CPUs utilized
perl5 function           dense  1259.29,msec,task-clock:u,1259292166,100.00,0.999,CPUs utilized
megaparsec-string        dense  2909.21,msec,task-clock:u,2909209205,100.00,1.000,CPUs utilized
megaparsec-bytestring    dense  3474.08,msec,task-clock:u,3474083010,100.00,1.000,CPUs utilized
megaparsec-text          dense  3886.78,msec,task-clock:u,3886779956,100.00,1.000,CPUs utilized
attoparsec-bytestring    dense  2511.68,msec,task-clock:u,2511684103,100.00,1.000,CPUs utilized
attoparsec-text          dense  3239.72,msec,task-clock:u,3239720725,100.00,0.999,CPUs utilized
attoparsec-text-lazy     dense  2610.54,msec,task-clock:u,2610541395,100.00,1.000,CPUs utilized
regex-applicative-string dense  11392.82,msec,task-clock:u,11392820150,100.00,1.000,CPUs utilized
pcre-heavy-text          dense  ∞
lens-regex-text          dense  ∞
lens-regex-bytestring    dense  ∞
stringsearch-bytestring  dense  694.62,msec,task-clock:u,694617573,100.00,0.996,CPUs utilized
sed                      sparse 3.01,msec,task-clock:u,3005735,100.00,0.823,CPUs utilized
python3                  sparse 54.20,msec,task-clock:u,54199590,100.00,0.991,CPUs utilized
python3 function         sparse 57.17,msec,task-clock:u,57172377,100.00,0.993,CPUs utilized
perl5                    sparse 1.92,msec,task-clock:u,1917385,100.00,0.705,CPUs utilized
perl5 function           sparse 3.96,msec,task-clock:u,3957148,100.00,0.828,CPUs utilized
megaparsec-string        sparse 248.56,msec,task-clock:u,248563674,100.00,0.999,CPUs utilized
megaparsec-bytestring    sparse 42.82,msec,task-clock:u,42823270,100.00,0.993,CPUs utilized
megaparsec-text          sparse 90.50,msec,task-clock:u,90497251,100.00,0.997,CPUs utilized
attoparsec-bytestring    sparse 28.05,msec,task-clock:u,28045324,100.00,0.989,CPUs utilized
attoparsec-text          sparse 29.34,msec,task-clock:u,29340344,100.00,0.989,CPUs utilized
attoparsec-text-lazy     sparse 21.91,msec,task-clock:u,21905039,100.00,0.987,CPUs utilized
regex-applicative-string sparse 415.87,msec,task-clock:u,415869898,100.00,0.999,CPUs utilized
pcre-heavy-text          sparse 9.77,msec,task-clock:u,9770756,100.00,0.969,CPUs utilized
lens-regex-text          sparse 9.35,msec,task-clock:u,9350947,100.00,0.968,CPUs utilized
lens-regex-bytestring    sparse 3.29,msec,task-clock:u,3285254,100.00,0.899,CPUs utilized
stringsearch-bytestring  sparse 1.44,msec,task-clock:u,1442014,100.00,0.843,CPUs utilized
```

That's more like it!

**Comment by expipiplus1** [comment]
I took the liberty of doing a minor version bump too.

**Comment by jamesdbrock** [comment]
I like the way this is shaping up, although I just tried to to run the tests with `cabal v2-test` and GHC 8.10 and compilation failed with

```
tests/TestTextLazy.hs:28:10: error:
    • Couldn't match expected type ‘T.Text’
                  with actual type ‘TS.Text’
```

How are you running the tests?

So you got lazy `streamEditT` working? That's great.

> Not sure about leaving all the other (now unused) functionality like `sepCap`, `findAll[Cap]`, `breakCap` and `splitCap` as these are all strict in the input.

Maybe we should not include those then?

**Comment by expipiplus1** [comment]
hmm, weird, seem to work fine here.

I've removed those functions.

File: src/Replace/Attoparsec/ByteString.hs [hunk 1]
```diff
@@ -227,7 +227,7 @@ streamEdit sep editor = runIdentity . streamEditT sep (Identity . editor)
 -- If you want the @editor@ function to remember some state,
 -- then run this in a stateful monad.
 streamEditT
-    :: (Monad m)
+    :: Applicative m
```
Comment (jamesdbrock) [inline src/Replace/Attoparsec/ByteString.hs:230]: Nice.

**Comment by jamesdbrock** [comment]
Thanks!

I will push a new package to Hackage.

I really appreciate your comments, especially the API documentation about how the laziness works in practice.

**Comment by jamesdbrock** [comment]
https://hackage.haskell.org/package/replace-attoparsec-1.4.5.0

**Comment by expipiplus1** [comment]
Thanks! Already using it!

**Comment by jamesdbrock** [comment]
I reworded your `anyTill` documentation a bit, please tell me if I got that wrong so we can fix it.

**Comment by expipiplus1** [comment]
> I reworded your `anyTill` documentation a bit, please tell me if I got that wrong so we can fix it.

LGTM!

**Pull request merged**